### PR TITLE
Removed the custom parser and used TrafficStats and fixed race condition with starting/stopping

### PR DESCRIPTION
**Pull request by SeyelentEco**
We didn’t need to use the custom Qtaguid parser since TrafficStats provided us all the info we needed.  

Also while testing I noticed a race condition with starting/stopping.  The MSG_STOP handler would cancel the MSG_START but it happened asynchronously.  If you called stopSampling() then startSampling() like in the sample app, it would schedule a MSG_STOP then a MSG_START, when MSG_STOP was handled, it would cancel the MSG_START accidentally.

### CODE REVIEW

File: connectionclass-sample/src/main/java/com/facebook/network/connectionclass/sample/MainActivity.java [hunk 1]
```diff
@@ -105,7 +106,10 @@ protected Void doInBackground(String... url) {
       String imageURL = url[0];
       try {
         // Open a stream to download the image from our URL.
-        InputStream input = new URL(imageURL).openStream();
+        URLConnection connection = new URL(imageURL).openConnection();
+        connection.setUseCaches(false);
```
Comment (jasta) [inline connectionclass-sample/src/main/java/com/facebook/network/connectionclass/sample/MainActivity.java]: Probably should be in a separate PR but whatev :)

File: connectionclass/src/main/java/com/facebook/network/connectionclass/DeviceBandwidthSampler.java [hunk 2]
```diff
@@ -81,13 +86,18 @@ public void startSampling() {
...
+      /**
+       * Resets previously read byte count after recording a sample, so that
+       * we don't count bytes downloaded in between sampling sessions.
+       */
+      sPreviousBytes = -1;
     }
   }
 
   private class SamplingHandler extends Handler {
     static final int MSG_START = 1;
```
Comment (jasta) [inline connectionclass/src/main/java/com/facebook/network/connectionclass/DeviceBandwidthSampler.java]: FWIW, the better pattern here is to make this private and add an explicit method to control the messages that go in (`sendSend` in this case).  That way, the sendStart method can do the necessary logic around removing the existing message.  In general, you want to keep the exact message semantics internal/private to the handler and expose a structured Java API only.

File: connectionclass/src/main/java/com/facebook/network/connectionclass/DeviceBandwidthSampler.java [hunk 3]
```diff
@@ -100,38 +110,28 @@ public void handleMessage(Message msg) {
...
-     * adding it to the BandwidthManager.
-     */
-    private void addSample() {
-      long byteDiff = QTagParser.getInstance().parseDataUsageForUidAndTag(Process.myUid());
+  /**
+   * Method for polling for the change in total bytes since last update and
+   * adding it to the BandwidthManager.
+   */
+  protected void addSample() {
+    long newBytes = TrafficStats.getTotalRxBytes();
```
Comment (jasta) [inline connectionclass/src/main/java/com/facebook/network/connectionclass/DeviceBandwidthSampler.java]: Good catch.  Didn't even realize this is exactly what we're doing now and it's also written in C by the platform.

File: connectionclass/src/main/java/com/facebook/network/connectionclass/DeviceBandwidthSampler.java [hunk 4]
```diff
@@ -100,38 +110,28 @@ public void handleMessage(Message msg) {
...
+    long byteDiff = newBytes - sPreviousBytes;
```
Comment (jasta) [inline connectionclass/src/main/java/com/facebook/network/connectionclass/DeviceBandwidthSampler.java]: `newBytes` itself can be `-1` and that should be tested.

File: connectionclass/src/main/java/com/facebook/network/connectionclass/DeviceBandwidthSampler.java [hunk 5]
```diff
@@ -10,16 +10,20 @@
...
+import android.os.SystemClock;
+import android.util.Log;
 
 import java.util.concurrent.atomic.AtomicInteger;
 
+import javax.annotation.Nonnull;
+
 /**
- * Class used to read from the file {@code /proc/net/xt_qtaguid/stats} periodically, in order to
+ * Class used to read from TrafficStats periodically, in order to
```
Comment (zlern2k) [inline connectionclass/src/main/java/com/facebook/network/connectionclass/DeviceBandwidthSampler.java]: nit: can probably merge these lines.

**Comment by jasta** [comment]
lgtm as-is, feel free to fold in my suggestions at your discretion.

**Comment by zlern2k** [comment]
Red diffs are ossm.

**Comment by zlern2k** [comment]
Alex -- give the word when you're ready for one of us to pull the merge trigger, or just go ahead and do it yourself once you've integrated feedback to your desired level.

**Comment by SeyelentEco** [comment]
Refactored the SamplingHandler to not leak any internal details and fixed a few nits.
